### PR TITLE
[SRVKS-324] Implement SMMR and NetworkPolicy behavior

### PR DESCRIPTION
**Pull request by nak3**
This patch migrates these PRs.

- https://github.com/openshift-knative/knative-openshift-ingress/pull/27
- https://github.com/openshift-knative/knative-openshift-ingress/pull/29
- https://github.com/openshift-knative/knative-openshift-ingress/pull/31

TODO:
- [x] Test on OCP clusters.
- [x] Add unit tests

### CODE REVIEW

File: pkg/reconciler/ingress/ingress.go [hunk 1]
```diff
@@ -175,6 +189,13 @@ func (r *Reconciler) reconcileIngress(ctx context.Context, ia *v1alpha1.Ingress)
 	ia.Status.InitializeConditions()
 	logger.Infof("Reconciling ingress: %#v", ia)
 
+	// Only add Istio ingress to SMMR
+	if ia.GetAnnotations()[networking.IngressClassAnnotationKey] == network.IstioIngressClassName {
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: Aren't we filtering by istio ingress in this component anyway currently?

File: pkg/reconciler/ingress/ingress.go [hunk 2]
```diff
@@ -324,10 +345,38 @@ func (r *Reconciler) reconcileRouteDeletion(ctx context.Context, ia *v1alpha1.In
...
 
+	ingresses, err := r.ingressLister.Ingresses(ia.GetNamespace()).List(labels.SelectorFromSet(map[string]string{}))
+	if err != nil {
+		return fmt.Errorf("failed to list ingress in %q: %w", ia.GetNamespace(), err)
+	}
+
+	// If the Ingress is the last one in the namespace, remove its namespace from SMMR.
+	if len(ingresses) == 1 {
+		// In order to double check that we are reconciling proper ingress check with name and namespace
+		if ia.GetNamespace() == ingresses[0].Namespace && ia.GetName() == ingresses[0].Name {
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: The namespace check is redundant, you can fold the name check into the first if, i.e.

```
if len(ingresses) == 1 && ia.GetName() == ingresses[0].Name {
```

File: pkg/reconciler/ingress/ingress.go [hunk 3]
```diff
@@ -623,3 +672,128 @@ func isRouteReady(r *routev1.Route) bool {
...
+		if err := r.deleteNetworkPolicy(ctx, ing, desired); err != nil {
+			return err
+		}
+	} else {
+		if err := r.ensureNetworkPolicy(ctx, ing, desired); err != nil {
+			return err
+		}
+	}
+
+	return nil
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: The entire `if` clause can be shortened to:

```golang
if isDeletion {
  return r.deleteNetworkPolicy(ctx, ing, desired)
}
return r.ensureNetworkPolicy(ctx, ing, desired)

File: pkg/reconciler/ingress/ingress.go [hunk 4]
```diff
@@ -623,3 +672,128 @@ func isRouteReady(r *routev1.Route) bool {
...
+}
+
+func (r *Reconciler) ensureNetworkPolicy(ctx context.Context, ing *v1alpha1.Ingress, desired *networkingv1.NetworkPolicy) error {
+	logger := logging.FromContext(ctx)
+
+	networkPolicy, err := r.GetKubeClient().NetworkingV1().NetworkPolicies(desired.Namespace).Get(desired.Name, metav1.GetOptions{})
+	if apierrs.IsNotFound(err) {
+		_, err = r.GetKubeClient().NetworkingV1().NetworkPolicies(desired.GetNamespace()).Create(desired)
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]:
```suggestion
		if _, err = r.GetKubeClient().NetworkingV1().NetworkPolicies(desired.GetNamespace()).Create(desired)
```

File: pkg/reconciler/ingress/ingress.go [hunk 5]
```diff
@@ -65,6 +71,12 @@ const (
 
 	notReadyOpenshiftIngressReason  = "OpenShiftIngressNotReady"
 	notReadyOpenshiftIngressMessage = "OpenShift Ingress is not ready"
+
+	// Namespace knative-serving-ingress hardcoded for now.
+	// The whole component knative-openshift-ingress is going to be moved into
+	// knative-serving-networking-openshift anyway, where it will be possible to statically determine the namespace to use.
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go:76]: We're doing that now 🤔 . But let's fix that in a separate PR.

File: pkg/reconciler/ingress/ingress.go [hunk 6]
```diff
@@ -623,3 +672,128 @@ func isRouteReady(r *routev1.Route) bool {
...
+			return members, false
+		}
+	}
+	return append(members, routeNamespace), true
+}
+
+func (r *Reconciler) reconcileNetworkPolicy(ctx context.Context, ing *v1alpha1.Ingress, isDeletion bool) error {
+	desired := oresources.MakeNetworkPolicyAllowAll(ing.GetNamespace())
+
+	networkPolices, err := r.GetKubeClient().NetworkingV1().NetworkPolicies(desired.Namespace).List(metav1.ListOptions{})
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]:
```suggestion
	networkPolicies, err := r.GetKubeClient().NetworkingV1().NetworkPolicies(desired.Namespace).List(metav1.ListOptions{})
```

**Review by markusthoemmes — COMMENTED** [review]
Just some flyby nits mostly.

File: pkg/reconciler/ingress/ingress.go [hunk 7]
```diff
@@ -175,6 +188,11 @@ func (r *Reconciler) reconcileIngress(ctx context.Context, ia *v1alpha1.Ingress)
 	ia.Status.InitializeConditions()
 	logger.Infof("Reconciling ingress: %#v", ia)
 
+	// Only add Istio ingress to SMMR
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: Comment no longer applies.

File: pkg/reconciler/ingress/ingress.go [hunk 8]
```diff
@@ -623,3 +666,120 @@ func isRouteReady(r *routev1.Route) bool {
...
+		return err
+	}
+
+	// update ServiceMeshMemberRole with the namespace info where knative routes created
+	smmr, err := r.smmrLister.ServiceMeshMemberRolls(smmrNamespace).Get(smmrName)
+	if err != nil {
+		return fmt.Errorf("failed to get SMMR in %q: %w", smmrNamespace, err)
+	}
+	newMembers, changed := appendIfAbsent(smmr.Spec.Members, ing.GetNamespace())
+	smmr.Spec.Members = newMembers
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: This modifies the informers copy (from the lister call above). We should DeepCopy it and only do that inside the `changed` bracket.

File: pkg/reconciler/ingress/ingress.go [hunk 9]
```diff
@@ -324,10 +341,36 @@ func (r *Reconciler) reconcileRouteDeletion(ctx context.Context, ia *v1alpha1.In
...
+		smmr, err := r.smmrLister.ServiceMeshMemberRolls(smmrNamespace).Get(smmrName)
+		if err != nil {
+			return fmt.Errorf("failed to get SMMR in %q: %w", smmrNamespace, err)
+		}
+		for i, val := range smmr.Spec.Members {
+			if val == ia.GetNamespace() {
+				smmr.Spec.Members = append(smmr.Spec.Members[:i], smmr.Spec.Members[i+1:]...)
+				break
+			}
+		}
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: Should we make a `removeIfPresent` function that does the inverse of `appendIfAbsent` and only do the update if something has really changed? Would make the code nicely symmetric.

File: hack/smmr-listers.patch [hunk 10]
```diff
@@ -0,0 +1,70 @@
+diff --git a/vendor/knative.dev/serving/pkg/reconciler/testing/v1alpha1/factory.go b/vendor/knative.dev/serving/pkg/reconciler/testing/v1alpha1/factory.go
```
Comment (markusthoemmes) [inline hack/smmr-listers.patch]: Should we make this one patch together with the serving listers? I assume that's easier to maintain as we'd only have to fix conflicts once, not twice.

File: pkg/reconciler/ingress/ingress_test.go [hunk 11]
```diff
@@ -469,6 +487,144 @@ func TestReconcile(t *testing.T) {
 			},
 			Name: "route-8a7e9a9d-fbc6-11e9-a88e-0261aff8d6d8-656566326438",
 		}},
+	}, {
+		Name: "create new NetworkPolicy",
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress_test.go]: Awesome tests, love 'em! ❤️

**Comment by nak3** [comment]
I found one minor issue and filed https://issues.redhat.com/browse/SRVKS-376. Except for it, it is working fine.

**Comment by nak3** [comment]
@markusthoemmes When you have time, can you please take a look?

File: pkg/reconciler/ingress/ingress.go [hunk 12]
```diff
@@ -324,10 +340,35 @@ func (r *Reconciler) reconcileRouteDeletion(ctx context.Context, ia *v1alpha1.In
 		return err
 	}
 
+	ingresses, err := r.ingressLister.Ingresses(ia.GetNamespace()).List(labels.SelectorFromSet(map[string]string{}))
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go]: Is the empty label selector necessary?

File: pkg/reconciler/ingress/ingress.go [hunk 13]
```diff
@@ -623,3 +664,130 @@ func isRouteReady(r *routev1.Route) bool {
...
+	if err != nil {
+		return fmt.Errorf("failed to get SMMR in %q: %w", smmrNamespace, err)
+	}
+	newMembers, changed := appendIfAbsent(smmr.Spec.Members, ing.GetNamespace())
+	if changed {
+		existing := smmr.DeepCopy()
+		existing.Spec.Members = newMembers
+		if _, err := r.smmrClient.MaistraV1().ServiceMeshMemberRolls(existing.Namespace).Update(existing); err != nil {
+			// ref for substring https://github.com/Maistra/istio-operator/blob/maistra-1.0/pkg/controller/servicemesh/validation/memberroll.go#L95
+			if strings.Contains(err.Error(), "one or more members are already defined in another ServiceMeshMemberRoll") {
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go:686]: I think it'd be super neat to surface this in the status of the Ingress to make it more actionable for the user. A potential seperate PR though imo.

File: pkg/reconciler/ingress/ingress_shift_test.go [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+	_ "github.com/openshift-knative/knative-serving-networking-openshift/pkg/client/maistra/injection/informers/maistra/v1/servicemeshmemberroll/fake"
+	_ "github.com/openshift-knative/knative-serving-networking-openshift/pkg/client/openshift/injection/informers/route/v1/route/fake"
+	oresources "github.com/openshift-knative/knative-serving-networking-openshift/pkg/reconciler/ingress/resources"
+
+	maistrav1 "github.com/maistra/istio-operator/pkg/apis/maistra/v1"
+)
+
+// TestReconcileShift is an additional tests of TestReconcile in ingress_test.go.
+// It tests extra resources like SMMR, NetworkPolicy for networking-openshift.
+func TestReconcileShift(t *testing.T) {
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress_shift_test.go]:
```suggestion
func TestReconcileOpenShift(t *testing.T) {
```

and rename the file to `ingress_openshift_test.go`?

**Review by markusthoemmes — COMMENTED** [review]
Great job! A few last words then this is done.

**Comment by markusthoemmes** [comment]
FWIW @nak3 the "first service takes long" is a known issue and even documented for the current k-o-i version. We should just make sure this doesn't make it worse.

File: pkg/reconciler/ingress/ingress.go (hunk 13 above)
Comment (nak3) [reply pkg/reconciler/ingress/ingress.go:686]: Yes, agreed. I opened the ticket for it https://issues.redhat.com/browse/SRVKS-377

**Comment by nak3** [comment]
Yes, it looks like "first service takes long" is the same issue with K-O-I.

I have addressed the last review comments.

File: pkg/reconciler/ingress/ingress.go [hunk 15]
```diff
@@ -324,10 +340,35 @@ func (r *Reconciler) reconcileRouteDeletion(ctx context.Context, ia *v1alpha1.In
 		return err
 	}
 
+	ingresses, err := r.ingressLister.Ingresses(ia.GetNamespace()).List(labels.Everything())
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/ingress.go:343]: Actually, on the label: Shall we filter this list by the `istio` class? We'd want to deal with the SMMR once we remove the last **istio** ingress, right?

Comment (nak3) [reply pkg/reconciler/ingress/ingress.go:343]: That makes sense. But the `istio` class (`networking.knative.dev/ingress.class`) is an annotation key not label key. I am wondering that it it difficult (impossible(?)) to use the annotation key.

**Pull request merged**